commands/previewPush.go

```diff
@@ -230,12 +264,119 @@ DomainLoop:
+
+				fmt.Printf("Removing domain from provider %s: %s\n", registrarName, domain)
+				*totalCorrections += 1
+				if domainRemover, ok := registrar.(providers.DomainRemover); ok && push {
+					err := domainRemover.EnsureDomainAbsent(domain)
+					if err != nil {
+						out.Errorf("Error deleting domain: %s\n", err)
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// DeleteUnmanagedZones iterates over all providers and their respective deployed zones, and removes those, which
+// are not configured within dnscontrol
+func DeleteUnmanagedZones(cfg *models.DNSConfig, createdProviders map[string]providers.DNSServiceProvider, push bool, out printer.CLI, totalCorrections *int) error {
+	for providerName, provider := range createdProviders {
+		out.StartUnmanagedZonesCheck(providerName)
+		zoneLister, ok := provider.(providers.ZoneLister)
+		if !ok {
+			out.Warnf("--purge-unmanaged-zones not implemented: provider %s\n", providerName)
+			continue
+		}
+		deployedZones, err := zoneLister.ListZones()
+		if err != nil {
+			return fmt.Errorf("failed ListZones for provider %s: %w\n", providerName, err)
+		}
+		for _, zone := range deployedZones {
+			if !IsZoneManagedByProvider(cfg, zone, providerName) {
+				fmt.Printf("Removing zone from provider %s: %s\n", providerName, zone)
+				*totalCorrections += 1
+				if zoneRemover, ok := provider.(providers.ZoneRemover); ok && push {
+					err := zoneRemover.EnsureZoneAbsent(zone)
+					if err != nil {
+						out.Errorf("Error deleting zone: %s\n", err)
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// GetDomainCfg returns the configuration defined for a domain within dnscontrol, or nil if the domain is not configured
+func GetDomainCfg(cfg *models.DNSConfig, domain string) *models.DomainConfig {
+	for _, domainCfg := range cfg.Domains {
+		if domainCfg.Name == domain {
+			return domainCfg
+		}
+	}
+	return nil
+}
+
+// IsDomainManagedByRegistrar checks whether `domain` is configured to be managed by `registrarName`
+func IsDomainManagedByRegistrar(cfg *models.DNSConfig, domain string, registrarName string) bool {
+	domainCfg := GetDomainCfg(cfg, domain)
+
+	// domain is not configured in dnscontrol at all
+	if domainCfg == nil {
+		return false
+	}
+	// domain is configured within dnscontrol, but with a different registrar
+	if domainCfg.RegistrarName != registrarName {
+		return false
+	}
+	return true
+}
+
+// IsZoneManagedByProvider checks whether `zone` is configured to be managed by `dnsProviderName
+func IsZoneManagedByProvider(cfg *models.DNSConfig, zone string, dnsProviderName string) bool {
+	domainCfg := GetDomainCfg(cfg, zone)
+
+	// domain is not configured in dnscontrol at all
+	if domainCfg == nil {
+		return false
+	}
+	_, ok := domainCfg.DNSProviderNames[dnsProviderName]
+	// domain is configured within dnscontrol, but not for this dns provider
+	if !ok {
+		return false
+	}
+	return true
+}
+
+type ProviderState struct {
+	createdRegistrars   map[string]providers.Registrar
```

Comment:
Use `instantiated` instead of `created`.

Reply:
I also thought about this one before... 
Naming this var `instantiated` would make me think the variable is one of  `RegistrarInstance` or `DNSProviderInstances`. That seemed confusing.